prost-derive/src/lib.rs

```diff
@@ -519,10 +519,52 @@ fn try_oneof(input: TokenStream) -> Result<TokenStream, Error> {
         }
     };
 
-    Ok(expanded.into())
+    Ok(expanded)
 }
 
 #[proc_macro_derive(Oneof, attributes(prost))]
-pub fn oneof(input: TokenStream) -> TokenStream {
-    try_oneof(input).unwrap()
+pub fn oneof(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    try_oneof(input.into()).unwrap().into()
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{try_message, try_oneof};
+    use quote::quote;
+
+    #[test]
+    fn test_rejects_colliding_message_fields() {
+        let output = try_message(quote!(
+            struct Invalid {
+                #[prost(bool, tag = "1")]
+                a: bool,
+                #[prost(oneof = "super::Whatever", tags = "4, 5, 1")]
+                b: Option<super::Whatever>,
+            }
+        ));
+        assert!(output.is_err());
+        assert_eq!(
+            output.unwrap_err().to_string(),
+            "message Invalid has multiple fields with tag 1"
+        );
+    }
+
+    #[test]
+    fn test_rejects_colliding_oneof_variants() {
+        let output = try_oneof(quote!(
+            pub enum Invalid {
+                #[prost(bool, tag = "1")]
+                A(bool),
+                #[prost(bool, tag = "3")]
+                B(bool),
+                #[prost(bool, tag = "1")]
+                C(bool),
+            }
+        ));
+        assert!(output.is_err());
+        assert_eq!(
+            output.unwrap_err().to_string(),
+            "invalid oneof Invalid: multiple variants have tag 1"
+        );
```

Comment:
The test for `variant with multiple tags` is missing.

Reply:
as far as i've been able to tell, that error case is unreachable by any input. any input that tries to put multiple "tag" attrs on a variant errors earlier, during the parsing of the attributes on the field.

i've added a test to demonstrate this here.

Reply:
Yes, that makes sense